runtime/cudaq/utils/tensor.h

```diff
@@ -0,0 +1,174 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <complex>
+#include <iterator>
+#include <vector>
+
+namespace cudaq {
+
+class matrix_2;
+
+matrix_2 operator*(const matrix_2 &, const matrix_2 &);
+matrix_2 operator*(std::complex<double>, const matrix_2 &);
+matrix_2 operator+(const matrix_2 &, const matrix_2 &);
+matrix_2 operator-(const matrix_2 &, const matrix_2 &);
+matrix_2 kronecker(const matrix_2 &, const matrix_2 &);
+template <typename Iterable,
+          typename T = typename std::iterator_traits<Iterable>::value_type>
+matrix_2 kronecker(Iterable begin, Iterable end);
+
+//===----------------------------------------------------------------------===//
+
+/// This is a minimalist matrix container. It is two-dimensional. It owns its
+/// data. Elements are of type `complex<double>`. Typically, it will contain a
+/// two-by-two set of values.
+class matrix_2 {
+public:
+  using Dimensions = std::pair<std::size_t, std::size_t>;
+
+  matrix_2() = default;
+  matrix_2(const matrix_2 &other)
+      : dimensions{other.dimensions},
+        data{new std::complex<double>[get_size(other.dimensions)]} {
+    std::copy(other.data, other.data + get_size(dimensions), data);
+  }
+  matrix_2(matrix_2 &&other) : dimensions{other.dimensions}, data{other.data} {
+    other.data = nullptr;
+  }
+  matrix_2(const std::vector<std::complex<double>> &v,
+           const Dimensions &dim = {2, 2})
+      : dimensions{dim}, data{new std::complex<double>[get_size(dim)]} {
+    assert(v.size() >= get_size(dimensions) &&
+           "vector must have enough elements");
+    std::copy(v.begin(), v.begin() + get_size(dimensions), data);
+  }
+  matrix_2(const std::complex<double> *v, const Dimensions &dim = {2, 2})
```

Comment:
Do you think we could use a `std::span<std::complex<double>>` instead of a pointer, so we can have boundary checks as well?

Reply:
span doesn't own data and I think we want ownership semantics from the meeting.

Reply:
The one case that one might get in trouble is building a matrix with a raw pointer to the data. I'm ok with that since I figure if one's using raw pointers, they need to know what they are doing.

Reply:
The normal case, using a vector, has an assertion.